### ISSUE TITLE
File path length validation (`NewPath`)

### ISSUE THREAD

**Issue by lmmx** [pydantic/pydantic#8308]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

There are limits on character length on file systems - for instance Linux is typically 255 but can be lower on `ecryptfs`-encrypted file systems - I have encountered this in libraries like HuggingFace Hub before and got it fixed so it can be parameterised from the file system programmatically (here https://github.com/huggingface/datasets/pull/3169):

```py
import os

os.statvfs(new_file_path).f_namemax
```

> ([Source](https://github.com/huggingface/datasets/pull/3173/files))

This would be a great addition to `NewPath`: I wasn't actually using it for validation but just had a very long filename (it originally came from an AI image generator whose image URL is the slugged text prompt) and when I saved it to disk my program crashed.

Of course I can check the filename length manually but:
- this assumes I'm thinking of this edge case at all
- since it is a quality of the file system then surely it comes under the same category of preconditions that we would want to be checked by `NewPath`

I took a look at the [docs](https://docs.pydantic.dev/latest/api/types/#pydantic.types.NewPath) for `NewPath` but it doesn't seem to have any parameters I can set to this end, however when I create one it does correctly error!

```
>>> TypeAdapter(NewPath).validate_python((("foo")*100+".txt"))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/louis/miniconda3/lib/python3.10/site-packages/pydantic/type_adapter.py", line 208, in validate_python
    return self.validator.validate_python(__object, strict=strict, from_attributes=from_attributes, context=context)
  File "/home/louis/miniconda3/lib/python3.10/site-packages/pydantic/types.py", line 606, in validate_new
    if path.exists():
  File "/home/louis/miniconda3/lib/python3.10/pathlib.py", line 1288, in exists
    self.stat()
  File "/home/louis/miniconda3/lib/python3.10/pathlib.py", line 1095, in stat
    return self._accessor.stat(self, follow_symlinks=follow_symlinks)
OSError: [Errno 36] File name too long: 'foofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoofoo.txt'
```

This is an `OSError` though, whereas we would surely want it to emit a `ValidationError`

```
>>> TypeAdapter(NewPath).validate_python("package.json")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/louis/miniconda3/lib/python3.10/site-packages/pydantic/type_adapter.py", line 208, in validate_python
    return self.validator.validate_python(__object, strict=strict, from_attributes=from_attributes, context=context)
pydantic_core._pydantic_core.ValidationError: 1 validation error for function-after[validate_new(), lax-or-strict[lax=union[json-or-python[json=function-after[path_validator(), str],python=is-instance[Path]],function-after[path_validator(), str]],strict=json-or-python[json=function-after[path_validator(), str],python=is-instance[Path]]]]
  Path already exists [type=path_exists, input_value='package.json', input_type=str]
```

From the source:

https://github.com/pydantic/pydantic/blob/20c0c6d9219e29a95f5e1cadeb05ec39d8c15c24/pydantic/types.py#L1212-L1218

what this does is

```
Path(("foo")*100+".txt").exists()
```

which will raise the OSError

Note that trying to `pathlib.Path.stat()` or `os.statvfs()` the path is at risk of raising this, whereas we can more safely expect that the parent directory to be `stat`-able. Note that pathlib's stat function doesn't include the `f_namemax` though.

```py
>>> p.parent.stat()
os.stat_result(st_mode=16832, st_ino=13238280, st_dev=58, st_nlink=53, st_uid=1000, st_gid=1000, st_size=16384, st_atime=1701773590, st_mtime=1701773589, st_ctime=1701773589)
>>> os.statvfs(p.parent)
os.statvfs_result(f_bsize=4096, f_frsize=4096, f_blocks=103473206, f_bfree=67819042, f_bavail=62544534, f_files=26353664, f_ffree=23318845, f_favail=23318845, f_flag=4102, f_namemax=143)
>>> os.statvfs(p.parent).f_namemax
143
```

So the safe way to check this in the routine then is:

```py
        if os.statvfs(path.parent).f_namemax < len(path.name):
            raise PydanticCustomError('path_too_long', 'Path name is too long')
        elif path.exists():
            raise PydanticCustomError('path_exists', 'Path already exists')
        elif not path.parent.exists():
            raise PydanticCustomError('parent_does_not_exist', 'Parent directory does not exist')
        else:
            return path
```

This way we check that it's safe to call `path.exists()`.

Note that this is only available on Unix ([docs](https://docs.python.org/3/library/os.html#os.statvfs)) so I would add a flavour check (I think an `isinstance` check is preferable to accessing `_flavour` directly).

```py
        if isinstance(p, pathlib.PosixPath) and os.statvfs(path.parent).f_namemax < len(path.name):
            ...
```

## Extension to other vfstat properties

While there, I considered if any other properties from vfstat ([docs](https://docs.python.org/3/library/os.html?highlight=f_flag#os.statvfs)) would be useful (ignoring any that'd be too involved).

In particular `f_flag` flags related to the filesystem, and of these the read-only flag also seems relevant to check (after the existence check, for safety).

```py
>>> os.statvfs(p.parent).f_flag & os.ST_RDONLY
0
```

## Suggestion

In short I suggest adding an import (`import os`) to `pydantic.types` and modifying its `PathType.validate_new` method to:

```py
        if isinstance(p, pathlib.PosixPath) and os.statvfs(path.parent).f_namemax < len(path.name):
            raise PydanticCustomError('path_too_long', 'Path name is too long')
        elif path.exists():
            raise PydanticCustomError('path_exists', 'Path already exists')
        elif not path.parent.exists():
            raise PydanticCustomError('parent_does_not_exist', 'Parent directory does not exist')
        elif isinstance(p, pathlib.PosixPath) and os.statvfs(path.parent).f_flag & os.ST_RDONLY:
            raise PydanticCustomError('path_readonly', 'Parent directory is read-only')
        else:
            return path
```

## Impact

This would potentially be more restrictive but given the logic is sound then I would expect it would shift left (i.e. find invalid paths earlier) and not actually introduce new exceptions. It would change OSErrors to ValidationErrors (via `PydanticCustomError`) but I think this is desirable: would welcome your feedback.

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [X] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by sydney-runkle** [comment]
@lmmx,

Should I add this issue to our 2.7.0 milestone? Would you be interested in working on this soon? 🚀

**Comment by lmmx** [comment]
Sure thing I will add it to my development kanban: Xmas break came and went without a chance for this ! Busy for next week then can sprint on this

**Comment by sydney-runkle** [comment]
@lmmx,

Sounds great! I've added this to the 2.7.0 milestone :).

**Comment by changhc** [comment]
Hi @sydney-runkle, I would like to work on this issue. After taking a brief look at the code, I have one question for you and @lmmx (thank you for the detailed issue description, by the way): how should this error be handled elsewhere?

This file name too long error might occur with other methods of `pathlib.Path` as well, like `.is_file()` and `is_dir()`. I have tested this locally. Now, these two methods are used to validate class `FilePath` and `DirectoryPath` respectively. Take `FilePath` for example. Currently we have
```py
def validate_file(path: Path, _: core_schema.ValidationInfo) -> Path:
    if path.is_file():
        return path
    else:
        raise PydanticCustomError('path_not_file', 'Path does not point to a file')
```

I'm thinking of changing this to
```py
def validate_file(path: Path, _: core_schema.ValidationInfo) -> Path:
    if isinstance(p, pathlib.PosixPath) and os.statvfs(path.parent).f_namemax < len(path.name):
        raise PydanticCustomError('path_not_file', 'Path does not point to a file')
    if path.is_file():
        return path
    else:
        raise PydanticCustomError('path_not_file', 'Path does not point to a file')
```

Basically catching the same case but raise a `path_not_file` error instead of a `path_too_long` error. The logic is that if a file name is too long, surely that file cannot exist. Showing the `path_too_long` error in this case is not helpful because `FilePath` is a path that must point to a file. Telling users that the path is too long is like not answering the question properly: "Okay, the path is too long, but then what? I want to know if this path actually leads to a file! Oh... if the path is too long, then the file cannot be there, right...?" and this can be a bit confusing. Therefore, I think it's better to raise a `path_not_file` error in this case. Same for `DirectoryPath`. Does this make sense?

**Comment by sydney-runkle** [comment]
Hi @changhc,

Great questions. I'm inclined to raise a `path_too_long` error, given the source of the issue. I definitely understand your logic though - perhaps to shine more light on the source of the problem, we could modify the message associated with such an error to include a note about the `f.namemax` restriction?

I'm also wondering - is there any chance this is going to be a breaking change for folks? I don't think so, but if so, I wonder if it makes more sense to document this as a custom validator example instead of implementing a logical change under the hood.

**Comment by sydney-runkle** [comment]
@lmmx, what are your thoughts on the above?

**Comment by lmmx** [comment]
Hey @sydney-runkle ooh this one fell off the task list, where were we 👾 By the way great talk at PyCon! 😄

You’re v welcome @changhc thanks for digging this back up!

Sure, extending to FilePath and any others is a good suggestion

## Error code options

My intuition when using NewPath is that a validation indicates “yes you can make this path”, so in that case I’d definitely want to indicate the most precise reason why not, and be constructive on how to solve, so `path_too_long` it is.

## Extension to `FilePath`

@changhc I hadn’t thought of extending scope here to `FilePath` but you’re right, it’s more precise to say that a file path doesn’t exist because the filename is too long to ever exist than to simply say “not found”. Yes I agree `name_too_long` is suitable here too.

## Breaking: for who?

I can imagine a case **in theory** which would exist where an impossible operation being invalidated would change code behaviour, it’d be a case where Pydantic was used to lint operations but not actually carry them out.

We could break this category in turn down into 2 kinds:

- those which actually carried out the operation (‘doomed’)
- those which never got around to the impossible operation (‘got away with it’)

We are talking about breaking the latter kind of program, which I think is likely negligible in the real world. (TODO: look at grep.app to gauge how this is used, at least in open source code)

Moreover, Pydantic model validation is **expected** to produce side effects (i.e. all validation might raise), so as long as we are not changing the type of the error (i.e. ValidationError to something else), which we’re not (either bringing OSError into ValidationError, or just changing its error code) then I would not call this a breaking change myself, no.

## Reproduction

I’m pretty excited by the new uv workspaces feature today like a lot of people, and just worked out a way to set up a PR demo here with a forked PR branch of the pydantic repo in a workspace package. I didn’t keep the work this structured last time which is part of why it didn’t get seen through to the finish line!

Repo is here, I’m fleshing it out with all of the above examples as TDD test cases now: https://github.com/lmmx/fp-len-validation-testing

**Comment by lmmx** [comment]
I tried putting the suggested block in `validate_new`

```py
        elif isinstance(path, PosixPath) and os.statvfs(path.parent).f_flag & os.ST_RDONLY:
            raise PydanticCustomError('path_readonly', 'Parent directory is read-only')
```

and made a demo read-only dir with `mkdir foo` then `chmod -R 333 foo`, to represent the parent of a new filepath

I'm not seeing it working, I'm going to take that one out and take the scope back to the file path length validation as per the title

**Comment by changhc** [comment]
@sydney-runkle Just to clarify, do you mean that you think we should raise `path_too_long` errors for `FilePath` and `DirectoryPath` as well?

I agree that for `NewPath` types (method `validate_new`) we should raise `path_too_long` errors. It's just that for `FilePath` (method `validate_file`) and `DirectoryPath` (method `validate_directory`) types I think it's better to "swallow" this `path_too_long` error and raise `path_not_file/directory` errors (@lmmx gets my point.)

About breaking someone's code, similar to what @lmmx mentioned, I think it's fine. Fixing this issue means we are bringing `OSError` into `ValidationError`. It will be a breaking change for someone who only handles `OSError` and lets the error get away but does not handle `ValidationError` -- the new `ValidationError` will be a surprise for them and maybe crash their program. However, I think this case is rather rare and I don't think it's that realistic. I assume that anyone using `FilePath`/`NewPath`/`DirectoryPath` is trying to validate paths. Then there might be
* folks not handling any error: they let the program fail anyway when validation fails. This change is not a breaking change for them since the new `ValidationError` will fail the program as well.
* folks expecting `ValidationError`: they want to deal with the error. As @lmmx mentioned, if they never hit this path too long error, this is not going to be a breaking change since they don't even know that `OSError` might be raised instead of `ValidationError`; if they hit the path too long `OSError` before, they would catch that and do something about it, but likely also terminate the program. For the latter group, this will change something but it should not be breaking since they will still get `ValidationError` and they are prepared for it.
* folks expecting `OSError` only: they know that `OSError` might be raised and just want to catch it somehow...? Looking at the current code, the only case that the model gives either the path or `OSError` is when one always has paths in an existing parent directory and one ensures that the paths never collide (no `path_exists` or `parent_does_not_exist` errors.) A possible use case might be when one allows users to specify file paths to store files but appends some suffix like timestamp to ensure that filenames don't collide. They would then deal with `OSError` only since otherwise all input values are valid. Changing `OSError` to `ValidationError` will break their error handling. (Sorry if this sounds confusing. I don't even believe this is an actual use case at all.)

All in all, I don't think this is going to be a big breaking change. It does change the validator's behaviour and users who are aware of the issue and already doing something about it will be affected. I think the impact is minor, but I don't know how these people are actually dealing with this.

**Comment by lmmx** [comment]
Looks like no major problems in the test (PR with just the initial suggested edit: #10207) besides what I assume is a `FileNotFoundError` being raised by `test_new_path_validation_parent_does_not_exist` in the test suite `tests/test_types.py`

```
    @staticmethod
    def validate_new(path: Path, _: core_schema.ValidationInfo) -> Path:
>       if isinstance(path, PosixPath) and os.statvfs(path.parent).f_namemax < len(path.name):
E       FileNotFoundError: [Errno 2] No such file or directory: '/nonexistentdir'
```

I resolved by updating to check the parent exists before os.statvfs'ing it.

**Comment by lmmx** [comment]
> @sydney-runkle Just to clarify, do you mean that you think we should raise `path_too_long` errors for `FilePath` and `DirectoryPath` as well?

Yeah I think Sydney is saying that it's reasonable to expect a user to reason through that "what does the file path too long have to do with anything...", I'm inclined to agree as it's more precise and gives the user a clearer idea of how to resolve it.

**Comment by lmmx** [comment]
OK I've now put in this name length check on the file and directory validator methods, there's a bit of repetition here but they are each distinct cases and don't want to DRY it out too much.

```diff
M: pydantic/types.py
@@ -1263,14 +1263,18 @@ class PathType:
 
     @staticmethod
     def validate_file(path: Path, _: core_schema.ValidationInfo) -> Path:
-        if path.is_file():
+        if isinstance(path, PosixPath) and os.statvfs(path.parent).f_namemax < len(path.name):
+            raise PydanticCustomError('path_too_long', 'Path name is too long')
+        elif path.is_file():
             return path
         else:
             raise PydanticCustomError('path_not_file', 'Path does not point to a file')
 
     @staticmethod
     def validate_directory(path: Path, _: core_schema.ValidationInfo) -> Path:
-        if path.is_dir():
+        if isinstance(path, PosixPath) and path.exists() and os.statvfs(path).f_namemax < len(path.name):
+            raise PydanticCustomError('path_too_long', 'Path name is too long')
+        elif path.is_dir():
             return path
         else:
             raise PydanticCustomError('path_not_directory', 'Path does not point to a directory')
```

To be clear, the parent directory of a file path is being checked in the file validator case here only to get the path length limit (which can change for different paths on a drive even, so you can't check anything 'in general', the parent directory is your only way to reliably tell).

**Comment by lmmx** [comment]
@changhc raises a good point: the new path could rely on parents (it's not our business how the user plans on making the new path)

> I guess we need to catch `OSError` in the validation methods anyway. The following will fail this check:
> 
> ```python
> >>> p = PosixPath("a"*1000 + "/" +"x"*1000)
> >>> p.parent.exists()
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/changhc/.pyenv/versions/3.12.3/lib/python3.12/pathlib.py", line 860, in exists
>     self.stat(follow_symlinks=follow_symlinks)
>   File "/home/changhc/.pyenv/versions/3.12.3/lib/python3.12/pathlib.py", line 840, in stat
>     return os.stat(self, follow_symlinks=follow_symlinks)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> OSError: [Errno 36] File name too long: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'
> ```
> 
> and this brings us back to the original problem. :(

- [via](https://github.com/pydantic/pydantic/pull/10207#discussion_r1725945368) PR review comment

I put in a new example / test case for this here: https://github.com/lmmx/fp-len-validation-testing/blob/master/examples/new_long_dir_and_file_names.py

**Comment by lmmx** [comment]
I've no idea what happened but somehow a bunch of datetime tests just broke :(

I can only think maybe it was the else raise part, but very confused now. Calling it a night, cannot repro on my local install

```
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃  File                    ┃  Function                                                          ┃  Function Line  ┃  Error Line  ┃  Error            ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│  tests/test_datetime.py  │  test_past_date_validation_fails[PastDate-value0]                  │  465            │  478         │  Failed           │
│  tests/test_datetime.py  │  test_past_date_validation_fails[past_date_type1-value0]           │  465            │  478         │  Failed           │
│  tests/test_datetime.py  │  test_future_date_validation_success[FutureDate-value0-result0]    │  490            │  501         │  ValidationError  │
│  tests/test_datetime.py  │  test_future_date_validation_success[future_date_type1-value0-re…  │  490            │  501         │  ValidationError  │
└──────────────────────────┴────────────────────────────────────────────────────────────────────┴─────────────────┴──────────────┴───────────────────┘
```

**Comment by eliottreich** [comment]
There's a $30 bounty for Pydantic validation edge cases on [Code.TaskBounty](https://code.task-bounty.com/task/fix-pydantic-validation-edge-case). Path validation improvements like this one are a good match.

Developers compete to implement it — you get working PRs to evaluate. Bounty is live: https://code.task-bounty.com/task/fix-pydantic-validation-edge-case
